test/unit/util_stream_spec.js

```diff
@@ -0,0 +1,386 @@
+        size(arr) {
+          return arr.length;
+        }
+      });
+
+      let reader = readable.getReader();
+      // Sleep for 10ms, so that read() is not unblocking the ready promise.
+      // Chain all read() to stream in sequence.
+      sleep(10).then(() => {
+        expect(log).toEqual('01');
+        return reader.read();
+      }).then((result) => {
+        expect(result.value).toEqual([1, 2, 3, 4]);
+        expect(result.done).toEqual(false);
+        return sleep(10);
+      }).then(() => {
+        expect(log).toEqual('01p2');
+        return reader.read();
+      }).then((result) => {
+        expect(result.value).toEqual([5, 6, 7, 8]);
+        expect(result.done).toEqual(false);
+        return sleep(10);
+      }).then(() => {
+        expect(log).toEqual('01p2p');
+        return reader.read();
+      }).then((result) => {
+        expect(result.value).toEqual(undefined);
+        expect(result.done).toEqual(true);
+        done();
+      });
+    });
+
+    it('should read data with blocking promise and buffer whole data' +
+       ' into stream', function (done) {
+      let log = '';
+      let port = new FakePort();
+      let messageHandler2 = new MessageHandler('worker', 'main', port);
+      messageHandler2.on('fakeHandler', (data, sink) => {
+        sink.onPull = function () {
+          log += 'p';
+        };
+        sink.onCancel = function (reason) {
+          log += 'c';
+        };
+        log += '0';
+        sink.ready.then(() => {
+          log += '1';
+          sink.enqueue([1, 2, 3, 4], 4);
+          return sink.ready;
+        }).then(() => {
+          log += '2';
+          sink.enqueue([5, 6, 7, 8], 4);
+          return sink.ready;
+        }).then(() => {
+          sink.close();
+        });
+        return sleep(10);
+      });
+
+      let messageHandler1 = new MessageHandler('main', 'worker', port);
+      let readable = messageHandler1.sendWithStream('fakeHandler', {}, {
+        highWaterMark: 8,
+        size(arr) {
+          return arr.length;
+        }
+      });
+
+      let reader = readable.getReader();
+
+      sleep(10).then(() => {
+        expect(log).toEqual('012');
+        return reader.read();
+      }).then((result) => {
+        expect(result.value).toEqual([1, 2, 3, 4]);
+        expect(result.done).toEqual(false);
+        return sleep(10);
+      }).then(() => {
+        expect(log).toEqual('012p');
+        return reader.read();
+      }).then((result) => {
+        expect(result.value).toEqual([5, 6, 7, 8]);
+        expect(result.done).toEqual(false);
+        return sleep(10);
+      }).then(() => {
+        expect(log).toEqual('012p');
+        return reader.read();
+      }).then((result) => {
+        expect(result.value).toEqual(undefined);
+        expect(result.done).toEqual(true);
+        done();
+      });
+    });
+
+    it('should ignore any pull after close is called', function (done) {
+      function createPromiseCapability() {
```

Comment:
Why are you duplicating this here, please import if from `shared/util.js` instead!